### PR TITLE
Media Store

### PR DESCRIPTION
**Pull request by tonyr59h**
TODO:
- [x] REST tests (@oguz)
- [x] Finish uploading for XMLRPC
- [x] XMLRPC tests (@tonyr59h)
- [x] Error handling (@tonyr59h)
- [x] Persistence tests (MediaStore unit tests) (@tonyr59h)
- [x] Update example app (@oguz)

### CODE REVIEW

**Comment by tonyr59h** [comment]
Tests completed! I'm glad @oguzkocer got to dig into the code while helping prepare this PR, thank you 🙇 !

This one's ready for review cc @maxme @aforcier

File: fluxc/build.gradle [hunk 1]
```diff
@@ -59,8 +59,8 @@ dependencies {
 
     // External libs
     compile 'org.greenrobot:eventbus:3.0.0'
-    compile 'com.squareup.okhttp3:okhttp:3.2.0'
-    compile 'com.squareup.okhttp3:okhttp-urlconnection:3.2.0'
+    compile 'com.squareup.okhttp3:okhttp:3.3.1'
+    compile 'com.squareup.okhttp3:okhttp-urlconnection:3.3.1'
```
Comment (maxme) [inline fluxc/build.gradle]: Are you using something specific from okhttp 3.3.+ or you just noticed it was outdated?

File: fluxc/src/main/java/org/wordpress/android/fluxc/utils/MediaUtils.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    private static final String MEDIA_CAPTION_KEY = "caption";
+    private static final String MEDIA_ALT_KEY = "alt";
+    private static final String MEDIA_PARENT_KEY = "parent_id";
+
+    /**
+     * Creates a {@link MediaModel} list from a WP.com REST response to a request for all media.
+     */
+    public static List<MediaModel> mediaListFromRestResponse(MultipleMediaResponse from) {
+        List<MediaModel> media = new ArrayList<>();
+        for (int i = 0; i < from.found; ++i) {
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/utils/MediaUtils.java]: I think this should be something like:
```java
for (int i = 0; i < from.media.size(); ++i) {
```

I get a OOB error on a site with a huge list of images (from.found = 1000+ while the from.media list is limited to 20)

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+    public class OnMediaProgress extends OnChanged {
+        public MediaModel media;
+        public float progress;
+        public OnMediaProgress(MediaModel media, float progress) {
+            this.media = media;
+            this.progress = progress;
+        }
+    }
+
+    public class OnMediaError extends OnChanged {
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: We recently changed how errors are managed https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/127 - We should drop that `OnMediaError` event and encapsulate errors in all other `OnChanged` events. 

See [OnSiteChanged](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/47417e5c9576bed50fab2f3000fc53c11f6c708e/fluxc/src/main/java/org/wordpress/android/fluxc/store/SiteStore.java#L95-L100) for example.

File: fluxc/build.gradle (hunk 1 above)
Comment (tonyr59h) [inline fluxc/build.gradle]: Just noticed it was outdated, I didn't specifically update it for anything. Think it should be reverted?

Comment (maxme) [inline fluxc/build.gradle]: Nope, I was just curious ;)

File: fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+                                AppLog.w(T.MEDIA, "could not parse Fetch media response, ID: " + media.getMediaId());
+                                MediaNetworkError error = MediaNetworkError.RESPONSE_PARSE_ERROR;
+                                notifyMediaError(MediaAction.FETCH_MEDIA, media, error);
+                            }
+                        }
+                    }, new BaseRequest.BaseErrorListener() {
+                @Override
+                public void onErrorResponse(@NonNull BaseRequest.BaseNetworkError error) {
+                    AppLog.v(T.MEDIA, "VolleyError Fetching media: " + error);
+                    notifyMediaError(MediaAction.FETCH_MEDIA, media, MediaNetworkError.UNKNOWN);
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java]: I think we want to "interpret" BaseNetworkError here and "convert" it into a MediaNetworkError.

File: fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+                        }
+                    }, new BaseRequest.BaseErrorListener() {
+                @Override
+                public void onErrorResponse(@NonNull BaseRequest.BaseNetworkError error) {
+                    AppLog.v(T.MEDIA, "VolleyError deleting media (ID=" + media.getMediaId() + "): " + error);
+                    if (error.type == BaseRequest.GenericErrorType.NOT_FOUND) {
+                        AppLog.i(T.MEDIA, "Attempted to delete media that does not exist remotely.");
+                        notifyMediaDeleted(MediaAction.DELETE_MEDIA, media);
+                    } else {
+                        notifyMediaError(MediaAction.DELETE_MEDIA, media, MediaNetworkError.UNKNOWN);
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java]: Same here

File: fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java (hunk 4 above)
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java]: For instance, if we get a `BaseNetworkError.NOT_FOUND`, we probably want to transform it into a `MediaNetworkError.MEDIA_NOT_FOUND`

File: fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+                                AppLog.w(T.MEDIA, "could not parse delete media response, ID: " + media.getMediaId());
+                                MediaNetworkError error = MediaNetworkError.RESPONSE_PARSE_ERROR;
+                                notifyMediaError(MediaAction.DELETE_MEDIA, media, error);
+                            }
+                        }
+                    }, new BaseRequest.BaseErrorListener() {
+                @Override
+                public void onErrorResponse(@NonNull BaseRequest.BaseNetworkError error) {
+                    AppLog.v(T.MEDIA, "VolleyError deleting media (ID=" + media.getMediaId() + "): " + error);
+                    if (error.type == BaseRequest.GenericErrorType.NOT_FOUND) {
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java]: (Same as here)

**Comment by maxme** [comment]
Using a device or emulator on Android 6.0+ - and don't give the example app permission to read images:

1. Sign in a self hosted site and try to upload a picture: you get a onMediaUpload event, with a GENERIC_ERROR. In that case it would be great to have a FS_PERMISSION_DENIED_TO_READ_FILE (or PERMISSION_DENIED_BY_FILESYSTEM). So the app could either ask the user for permission or notify the user it can't access the media library.

2. Sign in a wpcom site and try to upload a picture: you never get any onMediaUpload event. We should get a onMediaUpload in all cases (error or not), and follow the same model as above.

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+    public class OnMediaChanged extends OnChanged<MediaError> {
+        public MediaAction cause;
+        public List<MediaModel> media;
+        public OnMediaChanged(MediaAction cause, List<MediaModel> media) {
+            this.cause = cause;
+            this.media = media;
+        }
+    }
+
+    public class OnMediaUploaded extends OnChanged<MediaError> {
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: Can we add a boolean (`completed`) or a method (`isCompleted()`) to this event? I'm not sure this is very safe to compare float as in `progress >= 1.0f`

File: fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+                .encodedPath(xmlrpcUrl.getPath())
+                .username(site.getUsername())
+                .password(site.getPassword())
+                .build();
+        okhttp3.Request request = new okhttp3.Request.Builder()
+                .url(url)
+                .post(requestBody)
+                .build();
+
+        mOkHttpClient.newCall(request).enqueue(new Callback() {
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java]: I know it's a PITA to use volley here, but if we use the "raw" OkHttpClient, we also have to port everything that was made for volley like the custom UserAgent and Http Auth headers like [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/develop/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/BaseXMLRPCClient.java#L56-L57)

File: fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+                } else {
+                    AppLog.w(T.MEDIA, "error uploading media: " + response);
+                    notifyMediaError(MediaAction.UPLOAD_MEDIA, null, MediaNetworkError.UNKNOWN);
+                }
+            }
+
+            @Override
+            public void onFailure(Call call, IOException e) {
+                AppLog.w(T.MEDIA, "media upload failed: " + e);
+                notifyMediaError(MediaAction.UPLOAD_MEDIA, null, MediaNetworkError.UNKNOWN);
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java]: Propagating the timeout error would be great in that case. 

For example, simulate a network disconnection: try to upload a huge image (or throttle your network), then enable airplane mode - that will generate an "UNKNOWN" error, would be great to have a TIMEOUT error instead, so the app could show the user a specific message or ask them to retry.

**Comment by maxme** [comment]
Minor changes in the example app in 264b55e

File: fluxc/src/main/java/org/wordpress/android/fluxc/module/ReleaseNetworkModule.java [hunk 10]
```diff
@@ -145,6 +147,25 @@ public SiteXMLRPCClient provideSiteXMLRPCClient(Dispatcher dispatcher,
...
+                                                  @Named("regular") RequestQueue requestQueue,
+                                                  @Named("regular") OkHttpClient okHttpClient,
+                                                  AccessToken token, UserAgent userAgent) {
+        return new MediaRestClient(appContext, dispatcher, requestQueue, okHttpClient, token, userAgent);
+    }
+
+    @Singleton
+    @Provides
+    public MediaXMLRPCClient provideMediaXMLRPCClient(Dispatcher dispatcher,
+                                                      @Named("regular") RequestQueue requestQueue,
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/module/ReleaseNetworkModule.java]: You must use `@Named("custom-ssl")` here and on the next line to get the certificate "manager" (`memorizingTrustManager`) for XMLRPC (ie. selfhosted sites) requests.

File: fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java (hunk 8 above)
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java]: BTW, we should add a huge comment somewhere to explain why we're not using volley for uploads.

File: fluxc/src/main/java/org/wordpress/android/fluxc/module/ReleaseNetworkModule.java [hunk 11]
```diff
@@ -156,7 +177,6 @@ public AccountRestClient provideAccountRestClient(Context appContext, Dispatcher
     @Provides
     public PostRestClient providePostRestClient(Context appContext, Dispatcher dispatcher,
                                                 @Named("regular") RequestQueue requestQueue,
-                                                AppSecrets appSecrets,
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/module/ReleaseNetworkModule.java]: 👍

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java (hunk 7 above)
Comment (tonyr59h) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: Added in a5ce2ecbf5db593547043c13db526d688d4649e1

**Comment by tonyr59h** [comment]
> it would be great to have a FS_PERMISSION_DENIED_TO_READ_FILE

Added new error in 86ff78d919ef8357f97d902758713aa13210a11d and updated the example app.

Ran into a little issue: the Android lifecycle stops/starts the Activity when opening the media picker which de-registers the Activity from receiving `OnChanged` events. The upload returns with an error before `onStart` so the Activity receives no event.

Not a huge deal, just had to defer the upload request, but it got me wondering if sticky events would be a useful feature. Has it been considered?

**Comment by oguzkocer** [comment]
@tonyr59h: How are we handling the `user doesn't have permission to see the media library` error? I can't find a relevant `MediaStore.MediaErrorType` for it.

**More context:**
I am currently working on the integration and started with the `MediaGridFragment` and changing the current `ApiHelper.SyncMediaLibraryTask` to `MediaActionBuilder.newFetchAllMediaAction(payload)` dispatch and handling the response. In the previous implementation, we have an error type `ApiHelper.ErrorType.NO_UPLOAD_FILES_CAP` which is checked to see if the user has access to see the media library in the callback. _The name is pretty misleading, but it's going away anyway_ :fire: I am just trying to find the right kind of `MediaStore.MediaErrorType` for it.

**Comment by oguzkocer** [comment]
@tonyr59h It looks like `MediaActionBuilder.newFetchAllMediaAction(payload)` is getting `ALL` media, is there a plan for adding filters before this PR is merged in? Especially for pagination purposes, it'd be great to be able to send an offset and possibly even get it back in the event, so we know which section is fetched. What do you think?

**Comment by tonyr59h** [comment]
> How are we handling the user doesn't have permission to see the media library error? I can't find a relevant MediaStore.MediaErrorType for it.

I think `MediaErrorType.UNAUTHORIZED` makes sense in this case. Right now it's throwing a generic error, I'll update it to recognize the unauthorized response (403) and use the proper error.

**Comment by tonyr59h** [comment]
> is there a plan for adding filters before this PR is merged in? Especially for pagination purposes, it'd be great to be able to send an offset and possibly even get it back in the event, so we know which section is fetched. What do you think?

Thanks for pointing me to the `Filter` class in WPAndroid. Filter incoming :)

File: fluxc/src/main/java/org/wordpress/android/fluxc/utils/MediaUtils.java [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    // File operations
+    //
+
+    /**
+     * Queries filesystem to determine if a given file can be read.
+     */
+    public static boolean canReadFile(String filePath) {
+        if (filePath == null || TextUtils.isEmpty(filePath)) return false;
+        File file = new File(filePath);
+        return file.canRead();
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/utils/MediaUtils.java]: I'm not sure why, but on some emulators (geny 7.0) this call `file.canRead()` returns false. The file exist, the path is correct. Nothing printed in the logs. Permission error?

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java [hunk 13]
```diff
@@ -0,0 +1,515 @@
...
+            for (MediaModel media : payload.media) {
+                MediaSqlUtils.deleteMedia(media);
+            }
+        }
+
+        emitChange(onMediaChanged);
+    }
+
+    private boolean isWellFormedForUpload(@NonNull MediaModel media) {
+        return BaseUploadRequestBody.hasRequiredData(media) == null;
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: We should log the error returned by `hasRequiredData`, can be handy to debug.

**Comment by maxme** [comment]
This is still an issue:

https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/122#r77520777 - we should reuse the same User Agent, maybe something like:

```diff
--- a/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/BaseXMLRPCClient.java
+++ b/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/BaseXMLRPCClient.java
@@ -20,7 +20,7 @@ public class BaseXMLRPCClient {
     private SiteModel mSiteModel;
     private final RequestQueue mRequestQueue;
     protected final Dispatcher mDispatcher;
-    private UserAgent mUserAgent;
+    protected UserAgent mUserAgent;
     protected OnAuthFailedListener mOnAuthFailedListener;
     private HTTPAuthManager mHTTPAuthManager;
 
diff --git a/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java b/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java
index 6217e2d..560ac93 100644
--- a/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java
+++ b/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java
@@ -224,6 +224,7 @@ public class MediaXMLRPCClient extends BaseXMLRPCClient implements ProgressListe
         okhttp3.Request request = new okhttp3.Request.Builder()
                 .url(url)
                 .post(requestBody)
+                .addHeader("User-Agent", mUserAgent.toString())
                 .build();
 
         mOkHttpClient.newCall(request).enqueue(new Callback() {
```

For HTTP Auth, this will be a bit more tricky

**Comment by maxme** [comment]
When trying to upload a media to a site with HTTP Auth enabled, I get the following error:

```
W/WordPress-MEDIA( 4890): error uploading media: Response{protocol=http/1.1, code=401, message=Authorization Required, url=http://[login]:[password]@bh.wpmt.co/xmlrpc.php}
D/WordPress-API( 4890): Dispatching action: MediaAction-UPLOADED_MEDIA
```

File: fluxc/src/main/java/org/wordpress/android/fluxc/action/MediaAction.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    // Remote responses
+    @Action(payloadType = MediaListPayload.class)   PUSHED_MEDIA,
+    @Action(payloadType = ProgressPayload.class)    UPLOADED_MEDIA,
+    @Action(payloadType = MediaListPayload.class)   FETCHED_MEDIA,
+    @Action(payloadType = MediaListPayload.class)   DELETED_MEDIA,
+
+    // Local actions
+    @Action(payloadType = MediaListPayload.class)   UPDATE_MEDIA,
+    @Action(payloadType = MediaListPayload.class)   REMOVE_MEDIA,
+}
```
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/action/MediaAction.java]: Fixed in `feature/39-media-store-small-fixes`

File: fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java (hunk 13 above)
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java]: Added in `feature/39-media-store-small-fixes`

**Comment by maxme** [comment]
On a wpcom site (I haven't tested on a self hosted yet, but that could be different):

There is something wrong when I try to upload on a slow network (throttled connection), the progress shows 99% very quickly then is stuck at 99% for 15 secs, and then I get a timeout. Using Charles I see the connection active and uploading (not stale, so the timeout at 15s is weird). 

I'm not sure why the progress is up to 99% so quickly, I wonder if we measure disk read instead of network write.

**Comment by maxme** [comment]
Relevant logs:
```
D/WordPress-API: Dispatching action: MediaAction-UPLOAD_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 1.0
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
W/WordPress-MEDIA: media upload failed: java.net.SocketTimeoutException: timeout
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
```

File: fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java [hunk 15]
```diff
@@ -0,0 +1,476 @@
...
+
+    public MediaRestClient(Context appContext, Dispatcher dispatcher, RequestQueue requestQueue,
+                           OkHttpClient okClient, AccessToken accessToken, UserAgent userAgent) {
+        super(appContext, dispatcher, requestQueue, accessToken, userAgent);
+        mOkHttpClient = okClient;
+    }
+
+    @Override
+    public void onProgress(MediaModel media, float progress) {
+        notifyMediaProgress(media, Math.max(progress, 0.99f), null);
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java]: Should be `Math.min(progress, 0.99f)` here (that's why we only see 99% progress in the logs)

File: fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+    public MediaXMLRPCClient(Dispatcher dispatcher, RequestQueue requestQueue, OkHttpClient okClient,
+                             AccessToken accessToken, UserAgent userAgent,
+                             HTTPAuthManager httpAuthManager) {
+        super(dispatcher, requestQueue, accessToken, userAgent, httpAuthManager);
+        mOkHttpClient = okClient;
+    }
+
+    @Override
+    public void onProgress(MediaModel media, float progress) {
+        notifyMediaProgress(media, Math.max(0.99f, progress), null);
```
Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java]: Same here, should be `Math.min(progress, 0.99f)`.

File: fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java (hunk 15 above)
Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java]: That's an unfortunate error :( Do you want me to fix that in `feature/39-media-store-small-fixes` or do you want to send a separate PR for it?

Comment (maxme) [inline fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java]: Same branch `feature/39-media-store-small-fixes` will work

Comment (oguzkocer) [inline fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java]: Fixed in `feature/39-media-store-small-fixes`

File: fluxc/src/main/java/org/wordpress/android/fluxc/action/MediaAction.java (hunk 14 above)
Comment (tonyr59h) [inline fluxc/src/main/java/org/wordpress/android/fluxc/action/MediaAction.java]: > makes the payload/action association more obvious

That's what I was going for, but it does make more sense to stick with our code style 👍

File: fluxc/src/main/java/org/wordpress/android/fluxc/utils/MediaUtils.java (hunk 12 above)
Comment (tonyr59h) [inline fluxc/src/main/java/org/wordpress/android/fluxc/utils/MediaUtils.java]: IIRC 627f6676cd7d8761d39ef8144d49d12b3cb8f17f should have fixed that problem...I'll investigate further.

**Comment by oguzkocer** [comment]
As far as I can tell the permission error is the only thing holding back this PR. If I missed anything else let's list them here so we can handle them asap. Once this is merged in, we can focus on the integration part which is at least half-way done.

**Comment by tonyr59h** [comment]
> As far as I can tell the permission error is the only thing holding back this PR.

That's my take as well. @maxme did we miss anything?

**Comment by aforcier** [comment]
I'm getting two failing XML-RPC tests (REST tests all run fine).

##### ReleaseStack_MediaXmlRpcTest.testDeleteMediaThatExists:

I'm seeing:

```
WordPress-MEDIA: Media doesn't have required data: null
```

in the logs, but the upload/delete network requests are all going through and responding fine.

##### ReleaseStack_MediaXmlRpcTest.testPushNewMedia

I'm seeing:
```
junit.framework.AssertionFailedError: expected:<PUSHED_MEDIA> but was:<NOT_FOUND_ERROR>
```

**Comment by aforcier** [comment]
If I use the example app to upload an image to a self-hosted site, I get a timeout:
```
W/WordPress-MEDIA: media upload failed: java.net.SocketTimeoutException: timeout
```
which results in a `GENERIC_ERROR`, but the image was actually uploaded (and sniffing the request shows a successful response arriving).

I tried this with a different (smaller) image and it worked. Here's the image that failed: https://cloudup.com/c3-NtHuD-Gn

(The tests in my above comment still fail when I use an image that uploads correctly from the example app).

**Comment by aforcier** [comment]
This seems as good a place as any to mention it - let's decide on a naming style for test classes. With this PR, we'll have:

```
ReleaseStack_AccountTest
ReleaseStack_DiscoveryTest
ReleaseStack_MediaRestTest
ReleaseStack_MediaXmlRpcTest
ReleaseStack_PostTestWPCOM
ReleaseStack_PostTestXMLRPC
ReleaseStack_SiteTest
ReleaseStack_SiteTestWPCOM
```

and we'll also have (https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/186):

```
ReleaseStack_CommentTestWPCom
ReleaseStack_CommentTestXMLRPC
```

**Comment by tonyr59h** [comment]
> ReleaseStack_MediaXmlRpcTest.testPushNewMedia

This test is actually irrelevant now that the auto-upload behavior of the `PUSH_MEDIA` action was removed.

**Comment by tonyr59h** [comment]
> let's decide on a naming style for test classes

My vote is for `WPCom` semantics ☑️

**Comment by aforcier** [comment]
> My vote is for WPCom semantics ☑️
>
> So media tests would be ReleaseStack_MediaTestWPCom and ReleaseStack_MediaTestXMLRPC.

I applied the 'WPCom & XMLRPC' naming scheme to the rest of our tests in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/196.

**Comment by maxme** [comment]
:shipit: 🎉
